extensions/opc/src/opc.cpp

```diff
@@ -0,0 +1,461 @@
+  bReq.nodesToBrowse = UA_BrowseDescription_new();
+  bReq.nodesToBrowseSize = 1;
+
+  UA_NodeId_copy(&nodeId, &bReq.nodesToBrowse[0].nodeId);
+  bReq.nodesToBrowse[0].resultMask = UA_BROWSERESULTMASK_ALL;
+
+  UA_BrowseResponse bResp = UA_Client_Service_browse(client, bReq);
+
+  UA_BrowseRequest_deleteMembers(&bReq);
+
+  for(size_t i = 0; i < bResp.resultsSize; ++i) {
+    for(size_t j = 0; j < bResp.results[i].referencesSize; ++j) {
+      UA_ReferenceDescription *ref = &(bResp.results[i].references[j]);
+      if (cb(clientPtr, ref, basePath)) {
+        if (ref->nodeClass == UA_NODECLASS_VARIABLE || ref->nodeClass == UA_NODECLASS_OBJECT) {
+          std::string browsename((char *) ref->browseName.name.data, ref->browseName.name.length);
+          traverse(clientPtr, ref->nodeId.nodeId, cb, basePath + browsename);
+        }
+      } else {
+        UA_BrowseResponse_deleteMembers(&bResp);
+        return;
+      }
+    }
+  }
+
+  UA_BrowseResponse_deleteMembers(&bResp);
+};
+
+bool exists(ClientPtr& clientPtr, UA_NodeId nodeId) {
+  bool retval = false;
+  auto callback = [&retval](ClientPtr& clientPtr, const UA_ReferenceDescription *ref, const std::string& pat) -> bool {
+    retval = true;
+    return false;  // If any node is found, the given node exists, so traverse can be stopped
+  };
+  traverse(clientPtr, nodeId, callback);
+  return retval;
+};
+
+UA_StatusCode translateBrowsePathsToNodeIdsRequest(ClientPtr& clientPtr, const std::string& path, std::vector<UA_NodeId>& foundNodeIDs, const std::shared_ptr<core::logging::Logger>& logger) {
+  logger->log_trace("Trying to find node id for %s", path.c_str());
+
+  UA_Client *client = clientPtr.get();
+
+  auto tokens = utils::StringUtils::split(path, "/");
+  std::vector<UA_UInt32> ids;
+  for(size_t i = 0; i < tokens.size(); ++i) {
+    UA_UInt32 val = (i ==0 ) ? UA_NS0ID_ORGANIZES : UA_NS0ID_HASCOMPONENT;
+    ids.push_back(val);
+  }
+
+  UA_BrowsePath browsePath;
+  UA_BrowsePath_init(&browsePath);
+  browsePath.startingNode = UA_NODEID_NUMERIC(0, UA_NS0ID_OBJECTSFOLDER);
+
+  browsePath.relativePath.elements = (UA_RelativePathElement*)UA_Array_new(tokens.size(), &UA_TYPES[UA_TYPES_RELATIVEPATHELEMENT]);
+  browsePath.relativePath.elementsSize = tokens.size();
+
+  for(size_t i = 0; i < tokens.size(); ++i) {
+    UA_RelativePathElement *elem = &browsePath.relativePath.elements[i];
+    elem->referenceTypeId = UA_NODEID_NUMERIC(0, ids[i]);
+    elem->targetName = UA_QUALIFIEDNAME_ALLOC(0, tokens[i].c_str());
+  }
+
+  UA_TranslateBrowsePathsToNodeIdsRequest request;
+  UA_TranslateBrowsePathsToNodeIdsRequest_init(&request);
+  request.browsePaths = &browsePath;
+  request.browsePathsSize = 1;
+
+  UA_TranslateBrowsePathsToNodeIdsResponse response = UA_Client_Service_translateBrowsePathsToNodeIds(client, request);
+
+  if(response.resultsSize < 1) {
+    logger->log_warn("No node id in response for %s", path.c_str());
+    return UA_STATUSCODE_BADNODATAAVAILABLE;
```

Comment:
`UA_BrowsePath_deleteMembers(&browsePath);` is not called on this path, causing a leak on error. I suggest a utils::ScopeGuard here.

Reply:
Good catch, added scopeguard.